### PR TITLE
feat(metrics): Add  cardinality limiter to indexer [sns-1651]

### PR DESCRIPTION
**Pull request by untitaker**
Reopen of https://github.com/getsentry/sentry/pull/38302 to avoid notification spam

See #38257 and https://www.notion.so/sentry/Metrics-Dimensionality-Limiting-df010a6a6d4e467ca3c5c19230db862b#4966fb9c07fc4394b720ad161c99a096. This is just the glue code and configuration options for using the cardinality limiter in the indexer. The actual implementation is TBD. This is safe to merge because the stub implementation does not actually limit anything at all, so it should be fast enough to do synchronously for now

## rollout plan

- [x] https://github.com/getsentry/sentry/pull/38446
- [x] set options to nothing in prod
- [ ] merge + deploy this PR
- [ ] check prod metrics: redis should not be used
- [ ] https://github.com/getsentry/sentry/pull/38445
- [ ] check prod metrics: redis should still not be used
- [ ] run qe tests?
- [ ] get a redis cluster and configure it
- [ ] run use_quota on a separate thread
- [ ] set a quota
- [ ] get rid of writes limiter?
- [ ] stop indexing tag values

### CODE REVIEW

File: src/sentry/sentry_metrics/consumers/indexer/batch.py [hunk 1]
```diff
@@ -79,6 +98,29 @@ def extract_strings(self) -> Mapping[int, Set[str]]:
                 )
                 continue
 
+    @metrics.wraps("process_messages.limit_messages")
+    def limit_messages(self, keys_to_remove: Sequence[PartitionIdxOffset]) -> None:
```
Comment (untitaker) [inline src/sentry/sentry_metrics/consumers/indexer/batch.py:102]: continuing the batch refactor discussion at https://github.com/getsentry/sentry/pull/38302#discussion_r959891729 here -- the batch now no longer concerns itself with rate limiting, but see the other PR for details

File: src/sentry/utils/hashlib.py [hunk 2]
```diff
@@ -33,7 +33,7 @@ def hash_value(h, value):
             hash_value(h, item)
     elif isinstance(value, dict):
         h.update(b"\x05" + str(len(value)).encode("utf-8"))
-        for k, v in value.items():
+        for k, v in sorted(value.items()):
```
Comment (untitaker) [inline src/sentry/utils/hashlib.py:36]: see https://github.com/getsentry/sentry/pull/38302#discussion_r959624049

File: src/sentry/sentry_metrics/indexer/limiters/cardinality.py [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+    multiple use cases dynamically we just keep the mapping of rate limiters in
+    this factory.
+
+    [Copied from sentry.sentry_metrics.indexer.limiters.writes]
+    """
+
+    def __init__(self) -> None:
+        self.rate_limiters: MutableMapping[str, TimeseriesCardinalityLimiter] = {}
+
+    def get_ratelimiter(self, config: MetricsIngestConfiguration) -> TimeseriesCardinalityLimiter:
```
Comment (untitaker) [inline src/sentry/sentry_metrics/indexer/limiters/cardinality.py:136]: @fpacifici 
It is now namespaced the same way as the writes limiter, but I wonder whether we want to actually namespace this rate limiter per-indexer-backend? Since we are protecting clickhouse more than the indexer here.

**Review by untitaker — COMMENTED** [review]
adding some review comments to continue open discussions

File: src/sentry/sentry_metrics/configuration.py [hunk 4]
```diff
@@ -60,6 +61,7 @@ def get_ingest_config(
                 internal_metrics_tag="release-health",
                 writes_limiter_cluster_options=settings.SENTRY_METRICS_INDEXER_WRITES_LIMITER_OPTIONS,
                 writes_limiter_namespace=RELEASE_HEALTH_PG_NAMESPACE,
+                cardinality_limiter_cluster_options=settings.SENTRY_METRICS_INDEXER_CARDINALITY_LIMITER_OPTIONS,
```
Comment (fpacifici) [inline src/sentry/sentry_metrics/configuration.py:64]: Should we provide an empty dictionary here for release health ? We are not going to add this there.

File: src/sentry/sentry_metrics/consumers/indexer/batch.py [hunk 5]
```diff
@@ -53,17 +65,24 @@ def invalid_metric_tags(tags: Mapping[str, str]) -> Sequence[str]:
...
         self.use_case_id = use_case_id
         self.outer_message = outer_message
+        self.cardinality_limiter_state: Optional[CardinalityLimiterState[PartitionIdxOffset]] = None
 
-    @metrics.wraps("process_messages.parse_outer_message")
-    def extract_strings(self) -> Mapping[int, Set[str]]:
-        org_strings = defaultdict(set)
-
+    @metrics.wraps("process_messages.extract_messages")
+    def extract_messages(self) -> None:
```
Comment (fpacifici) [inline src/sentry/sentry_metrics/consumers/indexer/batch.py:83]: Since now this is the method that initializes the state of the batch.
Let's move it into the initializer (or create a static factory method if you prefer). 
So we remove the sequential coupling: init, then having to call `extract_messages` to do something correct. This does not return anything anymore so it is just initializing the valid state for this class.

File: src/sentry/sentry_metrics/consumers/indexer/batch.py (hunk 1 above)
Comment (fpacifici) [inline src/sentry/sentry_metrics/consumers/indexer/batch.py:102]: While we use this for rate limiting, its behavior is simply a filter. 
Just call it filter.

File: src/sentry/sentry_metrics/consumers/indexer/batch.py [hunk 6]
```diff
@@ -79,6 +98,29 @@ def extract_strings(self) -> Mapping[int, Set[str]]:
...
+                    },
+                )
+
+            self.skipped_offsets.add(offset)
+
+    @metrics.wraps("process_messages.extract_strings")
+    def extract_strings(self) -> Mapping[int, Set[str]]:
+        org_strings = defaultdict(set)
+
         for partition_offset, message in self.parsed_payloads_by_offset.items():
```
Comment (fpacifici) [inline src/sentry/sentry_metrics/consumers/indexer/batch.py:124]: I think you also need to skip the messages that you will have removed via `limit_messages` (or filter if we rename it). 
Otherwise we would be returning strings that we filtered through the cardinality limiter.

File: src/sentry/sentry_metrics/consumers/indexer/batch.py [hunk 7]
```diff
@@ -53,17 +65,24 @@ def invalid_metric_tags(tags: Mapping[str, str]) -> Sequence[str]:
...
 
+class InboundMessage(TypedDict):
+    # Note: This is only the subset of fields we access in this file.
+    org_id: int
+    name: str
+    type: str
+    tags: Mapping[str, str]
+
+
 class IndexerBatch:
```
Comment (fpacifici) [inline src/sentry/sentry_metrics/consumers/indexer/batch.py:76]: Please add a unit test for the logic you added to this class.
- set up a batch with messages
- filter it out with `limit_messages` 
- reconstruct messages and see the filtered one are skipped.
- extract strings and ensure the strings in the messages you filtered are not there.
(also let's see that if a string is present both in a message that is filtered out and in a message that is returned, the string is returned)

File: src/sentry/sentry_metrics/indexer/limiters/cardinality.py [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+    ) -> CardinalityLimiterState[TMessageKey]:
+        request_hashes = defaultdict(set)
+        keys_to_remove = {}
+        for key, message in messages.items():
+            org_id = message["org_id"]
+            message_hash = int(
+                hash_values(
+                    [
+                        message["name"],
+                        message["type"],
```
Comment (fpacifici) [inline src/sentry/sentry_metrics/indexer/limiters/cardinality.py:87]: I don't think we need `type` in the hash. If the same time series show up in two types (which are different tables in clickhouse) we can count only 1.

File: src/sentry/sentry_metrics/indexer/limiters/cardinality.py [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+
+class InboundMessage(TypedDict):
+    # Note: This is only the subset of fields we access in this file.
+    org_id: int
+    name: str
+    type: str
+    tags: Mapping[str, str]
+
+
+class TimeseriesCardinalityLimiter:
```
Comment (fpacifici) [inline src/sentry/sentry_metrics/indexer/limiters/cardinality.py:71]: Please add a unit test. There is enough logic in this class that is granted a test even if the rate limiter does not do anything.

File: src/sentry/sentry_metrics/indexer/limiters/cardinality.py (hunk 3 above)
Comment (fpacifici) [reply src/sentry/sentry_metrics/indexer/limiters/cardinality.py:137]: That is right. We do not need that, at least not in the immediate.

File: src/sentry/sentry_metrics/consumers/indexer/batch.py (hunk 1 above)
Comment (untitaker) [reply src/sentry/sentry_metrics/consumers/indexer/batch.py]: done

File: src/sentry/sentry_metrics/consumers/indexer/batch.py (hunk 6 above)
Comment (untitaker) [reply src/sentry/sentry_metrics/consumers/indexer/batch.py:126]: fixed

File: src/sentry/sentry_metrics/indexer/limiters/cardinality.py (hunk 3 above)
Comment (untitaker) [reply src/sentry/sentry_metrics/indexer/limiters/cardinality.py:136]: i tried to fix that by having a separate namespace config for the cardinality limiter.

File: src/sentry/sentry_metrics/indexer/limiters/cardinality.py (hunk 8 above)
Comment (untitaker) [reply src/sentry/sentry_metrics/indexer/limiters/cardinality.py]: that's fair, `type` is not a very large dimension

File: src/sentry/sentry_metrics/configuration.py (hunk 4 above)
Comment (untitaker) [reply src/sentry/sentry_metrics/configuration.py]: done

File: src/sentry/sentry_metrics/consumers/indexer/batch.py (hunk 5 above)
Comment (untitaker) [reply src/sentry/sentry_metrics/consumers/indexer/batch.py]: I made the method private and call it from init. Moving the code itself looked slightly uglier if i wanted to preserve the timer metric.

File: src/sentry/sentry_metrics/indexer/limiters/cardinality.py (hunk 9 above)
Comment (untitaker) [reply src/sentry/sentry_metrics/indexer/limiters/cardinality.py:71]: added some basic tests, found some pretty severe bugs 💯

File: src/sentry/sentry_metrics/consumers/indexer/batch.py (hunk 7 above)
Comment (untitaker) [reply src/sentry/sentry_metrics/consumers/indexer/batch.py:76]: done

File: src/sentry/sentry_metrics/consumers/indexer/batch.py [hunk 10]
```diff
@@ -79,8 +100,35 @@ def extract_strings(self) -> Mapping[int, Set[str]]:
...
 
+    @metrics.wraps("process_messages.filter_messages")
+    def filter_messages(self, keys_to_remove: Sequence[PartitionIdxOffset]) -> None:
+        for offset in keys_to_remove:
+            metrics.incr(
+                "sentry_metrics.indexer.process_messages.dropped_message",
+                tags={
+                    "reason": "cardinality_limit",
+                },
+            )
```
Comment (nikhars) [inline src/sentry/sentry_metrics/consumers/indexer/batch.py:111]: Instead of incrementing this in the loop why not just count the length of `keys_to_remove` and increment with that number once.

File: src/sentry/sentry_metrics/consumers/indexer/batch.py [hunk 11]
```diff
@@ -53,17 +65,26 @@ def invalid_metric_tags(tags: Mapping[str, str]) -> Sequence[str]:
...
+    name: str
+    type: str
+    tags: Mapping[str, str]
+
+
 class IndexerBatch:
     def __init__(self, use_case_id: UseCaseKey, outer_message: Message[MessageBatch]) -> None:
         self.use_case_id = use_case_id
         self.outer_message = outer_message
+        self.cardinality_limiter_state: Optional[CardinalityLimiterState[PartitionIdxOffset]] = None
```
Comment (fpacifici) [inline src/sentry/sentry_metrics/consumers/indexer/batch.py:80]: You are never using this.

File: src/sentry/sentry_metrics/indexer/limiters/cardinality.py [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+    def __init__(self, namespace: str, rate_limiter: CardinalityLimiter) -> None:
+        self.namespace = namespace
+        self.backend: CardinalityLimiter = rate_limiter
+
+    def check_cardinality_limits(
+        self, use_case_id: UseCaseKey, messages: Mapping[TMessageKey, InboundMessage]
+    ) -> CardinalityLimiterState[TMessageKey]:
```
Comment (fpacifici) [inline src/sentry/sentry_metrics/indexer/limiters/cardinality.py:79]: Why making this class generic with respect to the message key but explicit with respect to the InboundMessage type ?
Since this method needs to know the internals of `InboundMessage` and you cannot add a `hash` method to the typed dict, why not just making this class knowledgeable of `PartitionIdxOffset`? The generic there does not give us anything and makes the code slightly less readable.

File: src/sentry/sentry_metrics/indexer/limiters/cardinality.py [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+                hash_values(
+                    [
+                        message["name"],
+                        message["tags"],
+                    ]
+                ),
+                16,
+            )
+            prefix = _build_quota_key(self.namespace, org_id)
+            keys_to_remove[prefix, message_hash] = key
```
Comment (fpacifici) [inline src/sentry/sentry_metrics/indexer/limiters/cardinality.py:94]: I think this is hard to follow. This code creates a `keys_to_remove` dictionary at a point where we did not yet talk to the rate limiter. So at this point, these are not `keys_to_remove`.
If you do not want to generate a dict that contains the keys to remove here, just call it anything other than `keys_to_remove`

**Review by fpacifici — COMMENTED** [review]
It is good to me, please see Nikhar's comments.

File: src/sentry/sentry_metrics/configuration.py [hunk 14]
```diff
@@ -60,6 +62,8 @@ def get_ingest_config(
                 internal_metrics_tag="release-health",
                 writes_limiter_cluster_options=settings.SENTRY_METRICS_INDEXER_WRITES_LIMITER_OPTIONS,
                 writes_limiter_namespace=RELEASE_HEALTH_PG_NAMESPACE,
+                cardinality_limiter_cluster_options={},
+                cardinality_limiter_namespace=RELEASE_HEALTH_PG_NAMESPACE,
```
[the remark this answers is not in the document]
Comment (untitaker) [reply src/sentry/sentry_metrics/configuration.py:66]: I did that in response to a comment by @fpacifici that we don't need to namespace the cardinality limiter by indexer backend. Note that the "postgres" variant is configured for the cardinality limiter even for the cloudspanner config.

File: src/sentry/sentry_metrics/consumers/indexer/batch.py (hunk 10 above)
Comment (untitaker) [reply src/sentry/sentry_metrics/consumers/indexer/batch.py:111]: fixed, and also updated the set.add to be a set.update. i don't know what to do with the error log. we could choose to just not emit error logs, but tagging by organization could be a usecase in the future...

File: src/sentry/sentry_metrics/indexer/limiters/cardinality.py (hunk 12 above)
Comment (untitaker) [reply src/sentry/sentry_metrics/indexer/limiters/cardinality.py:79]: because i'm lazy and I don't want to construct valid PartitionIdxOffset in `test_cardinality_limiter.py` :D

but actually partitionidxoffset seems easier to construct than i remember so i'll do that

Comment (untitaker) [reply src/sentry/sentry_metrics/indexer/limiters/cardinality.py:79]: actually i remember, the real reason i did that was because of circular imports between `indexer.batch` and `indexer.limiters`, which are now gone due to another suggestion of yours 👍

File: src/sentry/sentry_metrics/indexer/limiters/cardinality.py (hunk 13 above)
Comment (untitaker) [reply src/sentry/sentry_metrics/indexer/limiters/cardinality.py:94]: ok, i fixed it by renaming, i am not sure if that's what you had in mind... it _does_ become `keys_to_remove` at some point though, so further down in the code the non-descript `hash_to_offset` is suboptimal naming too

**Review by nikhars — APPROVED** [review]
LGTM
